### ISSUE TITLE
Consumer Group - Quietly Fails to Read any Kafka Messages [v1.1.4]

### ISSUE THREAD

**Issue by Ferix9288** [twmb/franz-go#98]
Hey @twmb , 

There was an incident where unfortunately, franz-go Kafka client never read any of its designated topics, and there were no error or fatal logs reported.

Here are the logs that I do see (business sensitive details hidden):
```
[2021-10-20 00:53:39.429] [INFO] beginning to manage the group lifecycle; group: <consumer_group_name>
[2021-10-20 00:53:39.429] [INFO] joining group; group: <consumer_group_name>
[2021-10-20 00:53:39.437] [INFO] join returned MemberIDRequired, rejoining with response's MemberID; group: <consumer_group_name>, member_id: kgo-f7332df0-e4c7-49c0-bf88-dc72bd906db4
[2021-10-20 00:53:39.539] [INFO] producing to a new topic for the first time, fetching metadata to learn its partitions; topic: <topic_1>
[2021-10-20 00:53:39.539] [INFO] producing to a new topic for the first time, fetching metadata to learn its partitions; topic: <topic_2>
[2021-10-20 00:53:39.553] [INFO] done waiting for metadata for new topic; topic: <topic_2>
[2021-10-20 00:53:39.553] [INFO] done waiting for metadata for new topic; topic: <topic_1>
```

When setting up the consumer group (`g.manage` and `g.joinAndSync`), it did fail once due to a rebalance in progress but seemed to correctly recover and started the heartbeat monitoring.
```
[2021-10-20 00:54:16.303] [INFO] joined, balancing group; group: <consumer_group_name>, member_id: kgo-f7332df0-e4c7-49c0-bf88-dc72bd906db4, instance_id: <nil>, generation: 4, balance_protocol: cooperative-sticky, leader: true
[2021-10-20 00:54:16.303] [INFO] balancing group as leader
[2021-10-20 00:54:16.303] [INFO] balance group member; id: kgo-f7332df0-e4c7-49c0-bf88-dc72bd906db4, interests: interested topics: [<topic_2>], previously owned: 
[2021-10-20 00:54:16.304] [INFO] balanced; plan: kgo-f7332df0-e4c7-49c0-bf88-dc72bd906db4{<topic_2>[0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15]}
[2021-10-20 00:54:16.304] [INFO] syncing; group: <consumer_group_name>, protocol_type: consumer, protocol: cooperative-sticky
[2021-10-20 00:54:16.306] [INFO] synced; group: <consumer_group_name>, assigned: <topic_2>[0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15]
[2021-10-20 00:54:16.306] [INFO] new group session begun; group: <consumer_group_name>, added: map[<topic_2>:[0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15]], lost: map[]
[2021-10-20 00:54:16.306] [INFO] beginning heartbeat loop; group: <consumer_group_name>
[2021-10-20 00:54:16.306] [INFO] heartbeat errored; group: <consumer_group_name>, err: REBALANCE_IN_PROGRESS: The group is rebalancing, so a rejoin is needed.
[2021-10-20 00:54:16.306] [INFO] fetching offsets for added partitions; group: <consumer_group_name>, added: map[<topic_2>:[0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15]]
[2021-10-20 00:54:16.306] [INFO] cooperative consumer calling onRevoke at the end of a session even though no partitions were lost; group: <consumer_group_name>
[2021-10-20 00:54:16.307] [INFO] heartbeat loop complete; group: <consumer_group_name>, err: REBALANCE_IN_PROGRESS: The group is rebalancing, so a rejoin is needed.
[2021-10-20 00:54:16.307] [ERROR] fetch offsets failed due to context cancelation; group: <consumer_group_name>
[2021-10-20 00:54:16.307] [INFO] joining group; group: <consumer_group_name>
[2021-10-20 00:54:16.309] [INFO] joined, balancing group; group: <consumer_group_name>, member_id: kgo-f7332df0-e4c7-49c0-bf88-dc72bd906db4, instance_id: <nil>, generation: 5, balance_protocol: cooperative-sticky, leader: true
[2021-10-20 00:54:16.309] [INFO] balancing group as leader
[2021-10-20 00:54:16.309] [INFO] balance group member; id: kgo-f7332df0-e4c7-49c0-bf88-dc72bd906db4, interests: interested topics: [<topic_1> <topic_2>], previously owned: <topic_2>[0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15]
[2021-10-20 00:54:16.309] [INFO] balanced; plan: kgo-f7332df0-e4c7-49c0-bf88-dc72bd906db4{<topic_2>[0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15], <topic_1>[0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15]}
[2021-10-20 00:54:16.309] [INFO] syncing; group: <consumer_group_name>, protocol_type: consumer, protocol: cooperative-sticky
[2021-10-20 00:54:16.312] [INFO] synced; group: <consumer_group_name>, assigned: <topic_1>[0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15], <topic_2>[0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15]
[2021-10-20 00:54:16.312] [INFO] new group session begun; group: <consumer_group_name>, added: map[<topic_1>:[0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15]], lost: map[]
[2021-10-20 00:54:16.312] [INFO] beginning heartbeat loop; group: <consumer_group_name>
[2021-10-20 00:54:16.312] [INFO] fetching offsets for added partitions; group: <consumer_group_name>, added: map[<topic_1>:[0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15]]
[2021-10-20 00:54:16.312] [INFO] fetched committed offsets; group: <consumer_group_name>
```

However, not a single read happened for those topics. I know this because:
- We log whenever we receive a new Kafka message from the topic and saw none in the logs (code snippet below)
- We have metrics hook for `OnFetchBatchRead` -  this was never called 
- We also use `kowl` to monitor the kafka topics; new messages were definitely coming in

## The Exact Client Code Loop

```go
	for {
		select {
		case <-ctx.Done():
			c.Logger.Log(kgo.LogLevelInfo, fmt.Sprintf("context closed for kafka consumer with topics %v", c.Topics))
			return nil
		default:
		}
		fetches := c.client.PollFetches(ctx)
		if errs := fetches.Errors(); len(errs) > 0 {
			c.Logger.Log(kgo.LogLevelWarn, fmt.Sprintf("kafka fetch failed because of errs: %v", errs))
		}
		if fetches.IsClientClosed() {
			return nil
		}

		iter := fetches.RecordIter()
		for !iter.Done() {
			record := iter.Next()
			c.Logger.Log(kgo.LogLevelInfo, fmt.Sprintf("received kafka msg (topic %v, partition %v): %v", record.Topic, record.Partition, string(record.Value)))
```

## Hypothesis 

My speculation is that somehow, this `consumer loop` exited silently with `fetches.IsClientClosed` and therefore stopped polling and reading. 

## The Workaround

A simple restart seemed to fix the issue.

## Questions 

1. Other alternative hypothesis is that maybe the internal client code reached a deadlock and got stuck somewhere. Could this be a possibility?
2. If you agree it's most likely `IsClientClosed` exit loop, is it to be expected that there were no `errors` or warning signs indicating this in the logs? 
3. Also based on the comments, `fetches.Errors()` is supposed to include an injected `ErrClientClosed` but that was never written out. Are the comments wrong?
4. We're adding robustness to this loop where if we encountered that error, we simply reconnect by creating a new`franz-go` client. Does this seem like a correct workaround to you or is there unintended side-effects with re-creating clients?

## More Details 

- We're running our services via docker containers and for now, have two replicas
- For each instance, we create 2 `kgo.Client` - one responsible for producing [writing to topics] & another for consuming [reading from topics]
- I performed an upgrade via docker-compose down and docker-compose up to a new image
- Briefly glanced at our metrics and had quick sanity test, everything seemed working 
- However, turns out that one of the two services completely failed to consume its topic and wasn't discovered til much later since only happened with 1
  - Another interesting tidbit is that producing for that `instance` was fine; only `consuming` from topics was failing

**Comment by twmb** [comment]
Hi there,

As a long delayed update, I found a small bug in the patch while writing another feature (so, the bug was not actually encountered, but it's worth it to fix): https://github.com/twmb/franz-go/commit/02560c7

This is released in v1.3.4.

**Comment by Ferix9288** [comment]
Excellent, thanks for the update Travis!
